consensus/raft/raft.go

```diff
@@ -5,173 +5,239 @@ import (
-func (r *Raft) WaitForLeader(ctx context.Context) error {
-	// Using Raft observers panics on non-64 architectures.
-	// This is a work around
-	if sixtyfour {
-		return r.waitForLeader(ctx)
-	}
-	return r.waitForLeaderLegacy(ctx)
+// returns the folder path after creating it.
+// if folder is empty, it uses baseDir+Default.
+func makeDataFolder(baseDir, folder string) (string, error) {
+	if folder == "" {
+		folder = filepath.Join(baseDir, DefaultDataSubFolder)
+	}
+
+	err := os.MkdirAll(folder, 0700)
+	if err != nil {
+		return "", err
+	}
+	return folder, nil
 }
 
-func (r *Raft) waitForLeader(ctx context.Context) error {
-	obsCh := make(chan hashiraft.Observation, 1)
-	filter := func(o *hashiraft.Observation) bool {
-		switch o.Data.(type) {
-		case hashiraft.LeaderObservation:
-			return true
-		default:
-			return false
+// create Raft servers configuration
+func makeServerConf(peers []peer.ID) hraft.Configuration {
+	sm := make(map[string]struct{})
+
+	servers := make([]raft.Server, 0)
+	for _, pid := range peers {
+		p := peer.IDB58Encode(pid)
+		_, ok := sm[p]
+		if !ok { // avoid dups
+			sm[p] = struct{}{}
+			servers = append(servers, raft.Server{
+				Suffrage: raft.Voter,
+				ID:       raft.ServerID(p),
+				Address:  raft.ServerAddress(p),
+			})
+		}
+	}
+	return hraft.Configuration{
+		Servers: servers,
+	}
+}
+
+// diffConfigurations returns the serverIDs added and removed from
+// c2 in relation to c1.
+func diffConfigurations(
+	c1, c2 hraft.Configuration) (added, removed []hraft.ServerID) {
+	m1 := make(map[hraft.ServerID]struct{})
+	m2 := make(map[hraft.ServerID]struct{})
+	added = make([]hraft.ServerID, 0)
+	removed = make([]hraft.ServerID, 0)
+	for _, s := range c1.Servers {
+		m1[s.ID] = struct{}{}
+	}
+	for _, s := range c2.Servers {
+		m2[s.ID] = struct{}{}
+	}
+	for k, _ := range m1 {
+		_, ok := m2[k]
+		if !ok {
+			removed = append(removed, k)
+		}
+	}
+	for k, _ := range m2 {
+		_, ok := m1[k]
+		if !ok {
+			added = append(added, k)
 		}
 	}
-	observer := hashiraft.NewObserver(obsCh, false, filter)
-	r.raft.RegisterObserver(observer)
-	defer r.raft.DeregisterObserver(observer)
-	ticker := time.NewTicker(time.Second)
+	return
+}
+
+// WaitForLeader holds until Raft says we have a leader.
+// Returns uf ctx is cancelled.
+func (rw *raftWrapper) WaitForLeader(ctx context.Context) (string, error) {
+	obsCh := make(chan hraft.Observation, 1)
+	if sixtyfour { // 32-bit systems don't support observers
+		observer := hraft.NewObserver(obsCh, false, nil)
+		rw.raft.RegisterObserver(observer)
+		defer rw.raft.DeregisterObserver(observer)
+	}
+	ticker := time.NewTicker(time.Second / 2)
 	for {
 		select {
 		case obs := <-obsCh:
-			switch obs.Data.(type) {
-			case hashiraft.LeaderObservation:
-				leaderObs := obs.Data.(hashiraft.LeaderObservation)
-				logger.Infof("Raft Leader elected: %s", leaderObs.Leader)
-				return nil
-			}
+			_ = obs
+			// See https://github.com/hashicorp/raft/issues/254
```

Comment:
Was it? I don't see it...

Reply:
I'm going off of looking for the commit in the history of master when it was first published in the month of June and not seeing it anywhere, I didn't see a specific revert.  Either way it's not in the binary release yet (actually now I'm doubting that this matters because we build from source right?  Or does gx freeze the version?).  I think I'll make an issue out of this so we don't forget to update when the change is live.

Reply:
we use a frozen version currently at the 1.0.0 tag. But yeah, I think there is something weird in that git history. It took me forever to find where the change came from.